File: apps/worker/worker/job_descriptions/compiler.py
import json
import re
import unicodedata
from collections.abc import Mapping, Sequence
from dataclasses import dataclass
from hashlib import blake2b
from typing import cast

from ..documents.vocabulary import load_vocabulary, mentioned_skills
from .schemas import (
	JOB_REQUIREMENTS_COMPILER_VERSION,
	JOB_REQUIREMENTS_PROMPT_VERSION,
	JOB_REQUIREMENTS_SCHEMA_VERSION,
	Assessability,
	CriterionType,
	ModelRequirementExtraction,
	PredicateOperator,
	RequirementCategory,
	SourceModality,
	SuggestedKind,
)

MAX_DRAFT_REQUIREMENTS = 50

HEADING_SECTIONS = {
	"requirements": "requirements",
	"qualifications": "requirements",
	"minimum qualifications": "requirements",
	"what you bring": "requirements",
	"what we are looking for": "requirements",
	"what we're looking for": "requirements",
	"preferred qualifications": "preferred",
	"preferred skills": "preferred",
	"nice to have": "preferred",
	"nice-to-have": "preferred",
	"bonus points": "preferred",
	"responsibilities": "responsibilities",
	"what you will do": "responsibilities",
	"what you'll do": "responsibilities",
	"about us": "about",
	"about the company": "about",
	"benefits": "benefits",
	"what we offer": "benefits",
	"equal opportunity": "legal",
	"how to apply": "application",
}

REQUIRED_CUE = re.compile(
	r"\b(must|required|minimum|need to|needs to|shall|essential)\b",
	re.IGNORECASE,
)
PREFERRED_CUE = re.compile(
	r"\b(preferred|nice[ -]to[ -]have|bonus|plus|desirable|ideally)\b",
	re.IGNORECASE,
)
YEARS_PATTERN = re.compile(r"\b(\d{1,2})\s*\+?\s*(?:years?|yrs?)\b", re.IGNORECASE)
DEGREE_LEVELS = {
	"doctorate": ("phd", "doctorate", "doctoral", "dphil"),
	"master": ("master", "mba", "mtech", "m.tech", "msc", "m.sc"),
	"bachelor": ("bachelor", "undergraduate degree", "btech", "b.tech", "bsc", "b.sc"),
}
PROHIBITED_PATTERN = re.compile(
	r"\b(age|gender|sex|race|ethnicity|religion|marital status|pregnan(?:t|cy)|disability)\b",
	re.IGNORECASE,
)
ATTESTATION_PATTERNS = {
	RequirementCategory.WORK_AUTHORIZATION: re.compile(
		r"\b(work authorization|authorized to work|visa sponsorship|right to work)\b",
		re.IGNORECASE,
	),
	RequirementCategory.LOCATION: re.compile(
		r"\b(relocat(?:e|ion)|reside in|based in|commutable distance)\b",
		re.IGNORECASE,
	),
	RequirementCategory.SCHEDULE: re.compile(
		r"\b(night shift|weekends?|on[ -]call|travel|time zone|working hours)\b",
		re.IGNORECASE,
	),
}
SOFT_SKILL_PATTERN = re.compile(
	r"\b(communication|team player|self[ -]starter|leadership style|interpersonal)\b",
	re.IGNORECASE,
)
CERTIFICATION_PATTERN = re.compile(
	r"\b(certification|certified|licen[cs]e|credential)\b",
	re.IGNORECASE,
)
BULLET_PATTERN = re.compile(r"^\s*(?:[-*•▪◦]+|\d+[.)])\s+")


@dataclass(frozen=True)
class JobBlock:
	id: str
	text: str
	section: str
	start_offset: int
	end_offset: int
	was_bullet: bool


def source_blocks(source_text: str) -> list[JobBlock]:
	blocks: list[JobBlock] = []
	section = "unknown"
	offset = 0
	for raw_line in source_text.splitlines(keepends=True):
		line = raw_line.rstrip("\r\n")
		stripped = line.strip()
		line_start = offset
		offset += len(raw_line)
		if not stripped:
			continue
		heading_key = re.sub(r"[:\s]+$", "", stripped).casefold()
		if heading_key in HEADING_SECTIONS and len(stripped) <= 80:
			section = HEADING_SECTIONS[heading_key]
			continue
		bullet = BULLET_PATTERN.match(line)
		content_start = bullet.end() if bullet else len(line) - len(line.lstrip())
		content = line[content_start:].strip()
		if len(content) < 3:
			continue
		local_start = line.find(content, content_start)
		start = line_start + local_start
		blocks.append(
			JobBlock(
				id=f"jd-b{len(blocks) + 1}",
				text=content,
				section=section,
				start_offset=start,
				end_offset=start + len(content),
				was_bullet=bullet is not None,
			)
		)
	if not source_text.endswith(("\n", "\r")) and offset < len(source_text):
		offset = len(source_text)
	return blocks


def blocks_for_model(blocks: Sequence[JobBlock]) -> list[dict[str, object]]:
	return [
		{"id": block.id, "section": block.section, "text": block.text}
		for block in blocks
	]


def compile_job_description(
	source_text: str,
	model_output: Mapping[str, object] | None = None,
	*,
	degraded: bool = False,
	degraded_reason: str | None = None,
) -> dict[str, object]:
	blocks = source_blocks(source_text)
	warnings: list[str] = []
	candidates = deterministic_candidates(blocks, warnings)
	if model_output is not None:
		model_candidates, model_warnings = grounded_model_candidates(model_output, blocks)
		warnings.extend(model_warnings)
		candidates.extend(model_candidates)
	requirements = deduplicate(candidates, warnings)[:MAX_DRAFT_REQUIREMENTS]
	if len(candidates) > MAX_DRAFT_REQUIREMENTS:
		warnings.append("Low-confidence requirements were omitted after the review limit")
	if degraded_reason:
		warnings.append(degraded_reason)
	quality_state = "ready" if requirements else "review_required"
	if not requirements:
		warnings.append("No explicit job requirements were found; add criteria manually")
	return {
		"schemaVersion": JOB_REQUIREMENTS_SCHEMA_VERSION,
		"compilerVersion": JOB_REQUIREMENTS_COMPILER_VERSION,
		"promptVersion": JOB_REQUIREMENTS_PROMPT_VERSION,
		"degraded": degraded,
		"qualityState": quality_state,
		"warnings": unique_strings(warnings),
		"requirements": requirements,
	}


def deterministic_candidates(
	blocks: Sequence[JobBlock], warnings: list[str]
) -> list[dict[str, object]]:
	candidates: list[dict[str, object]] = []
	for block in blocks:
		text = normalized_statement(block.text)
		if block.section in {"about", "benefits", "legal", "application"}:
			continue
		if PROHIBITED_PATTERN.search(text):
			warnings.append(f"Potentially prohibited criterion omitted from {block.id}")
			continue
		skills = sorted(mentioned_skills(text), key=str.casefold)
		has_cue = REQUIRED_CUE.search(text) or PREFERRED_CUE.search(text)
		in_requirement_section = block.section in {"requirements", "preferred"}
		if block.section == "responsibilities" and not has_cue:
			continue
		if not in_requirement_section and not has_cue and not (block.was_bullet and skills):
			continue
		candidate = deterministic_candidate(block, text, skills)
		candidates.append(candidate)
	return candidates


def deterministic_candidate(
	block: JobBlock, text: str, skills: Sequence[str]
) -> dict[str, object]:
	suggested_kind, source_modality = infer_importance(block.section, text)
	category, assessability = infer_category_and_assessability(text, skills)
	predicate = infer_predicate(text, skills)
	evidence = [
		{
			"blockId": block.id,
			"quote": block.text,
			"startOffset": block.start_offset,
			"endOffset": block.end_offset,
			"section": block.section,
		}
	]
	signals = ["section"] if block.section in {"requirements", "preferred"} else []
	if skills:
		signals.append("taxonomy")
	if REQUIRED_CUE.search(text) or PREFERRED_CUE.search(text):
		signals.append("language_cue")
	confidence = min(0.92, 0.55 + 0.12 * len(signals))
	return finalize_candidate(
		{
			"normalizedText": text,
			"category": category.value,
			"suggestedKind": suggested_kind.value,
			"suggestedWeight": 2 if suggested_kind is SuggestedKind.REQUIRED else 1,
			"sourceModality": source_modality.value,
			"assessability": assessability.value,
			"predicate": predicate,
			"evidence": evidence,
			"confidence": confidence,
			"signals": signals or ["bullet"],
		}
	)


def infer_importance(section: str, text: str) -> tuple[SuggestedKind, SourceModality]:
	if PREFERRED_CUE.search(text):
		return SuggestedKind.PREFERRED, SourceModality.EXPLICIT_PREFERRED
	if REQUIRED_CUE.search(text):
		return SuggestedKind.REQUIRED, SourceModality.EXPLICIT_REQUIRED
	if section == "preferred":
		return SuggestedKind.PREFERRED, SourceModality.SECTION_PREFERRED
	if section == "requirements":
		return SuggestedKind.REQUIRED, SourceModality.SECTION_REQUIRED
	return SuggestedKind.PREFERRED, SourceModality.UNCLEAR


def infer_category_and_assessability(
	text: str, skills: Sequence[str]
) -> tuple[RequirementCategory, Assessability]:
	for category, pattern in ATTESTATION_PATTERNS.items():
		if pattern.search(text):
			return category, Assessability.CANDIDATE_ATTESTATION
	if SOFT_SKILL_PATTERN.search(text):
		return RequirementCategory.SOFT_SKILL, Assessability.RECRUITER_REVIEW
	if YEARS_PATTERN.search(text):
		return RequirementCategory.EXPERIENCE, Assessability.RESUME_EVIDENCE
	if education_level(text):
		return RequirementCategory.EDUCATION, Assessability.RESUME_EVIDENCE
	if CERTIFICATION_PATTERN.search(text):
		return RequirementCategory.CERTIFICATION, Assessability.RESUME_EVIDENCE
	if skills:
		return RequirementCategory.SKILL, Assessability.RESUME_EVIDENCE
	return RequirementCategory.OTHER, Assessability.UNCLEAR


def infer_predicate(text: str, skills: Sequence[str]) -> dict[str, object]:
	criteria: list[dict[str, object]] = []
	years = YEARS_PATTERN.search(text)
	if years:
		criteria.append(
			{
				"type": CriterionType.EXPERIENCE.value,
				"canonicalName": None,
				"minimumMonths": int(years.group(1)) * 12,
				"minimumLevel": None,
				"subjects": list(skills),
			}
		)
	elif skills:
		for skill in skills:
			criteria.append(
				{
					"type": CriterionType.SKILL.value,
					"canonicalName": skill,
					"minimumMonths": None,
					"minimumLevel": None,
					"subjects": [],
				}
			)
	level = education_level(text)
	if level:
		criteria.append(
			{
				"type": CriterionType.EDUCATION.value,
				"canonicalName": None,
				"minimumMonths": None,
				"minimumLevel": level,
				"subjects": [],
			}
		)
	if CERTIFICATION_PATTERN.search(text) and not criteria:
		criteria.append(
			{
				"type": CriterionType.CERTIFICATION.value,
				"canonicalName": text,
				"minimumMonths": None,
				"minimumLevel": None,
				"subjects": [],
			}
		)
	if not criteria:
		criteria.append(
			{
				"type": CriterionType.OTHER.value,
				"canonicalName": None,
				"minimumMonths": None,
				"minimumLevel": None,
				"subjects": [],
			}
		)
	operator = (
		PredicateOperator.ANY_OF
		if re.search(r"\bor\b", text, re.IGNORECASE)
		else PredicateOperator.ALL_OF
	)
	return {"operator": operator.value, "criteria": criteria}


def grounded_model_candidates(
	model_output: Mapping[str, object], blocks: Sequence[JobBlock]
) -> tuple[list[dict[str, object]], list[str]]:
	try:
		extraction = ModelRequirementExtraction.model_validate(model_output)
	except ValueError:
		return [], ["Model requirement output failed schema validation"]
	by_id = {block.id: block for block in blocks}
	candidates: list[dict[str, object]] = []
	dropped = 0
	for requirement in extraction.requirements:
		evidence: list[dict[str, object]] = []
		for citation in requirement.evidence:
			block = by_id.get(citation.block_id)
			if block is None:
				continue
			local_start = block.text.find(citation.quote)
			if local_start < 0:
				continue
			evidence.append(
				{
					"blockId": block.id,
					"quote": citation.quote,
					"startOffset": block.start_offset + local_start,
					"endOffset": block.start_offset + local_start + len(citation.quote),
					"section": block.section,
				}
			)
		if not evidence:
			dropped += 1
			continue
		if requirement.assessability is Assessability.PROHIBITED:
			dropped += 1
			continue
		predicate = normalize_model_predicate(
			requirement.predicate.model_dump(by_alias=True)
		)
		kind = requirement.suggested_kind
		candidate = finalize_candidate(
			{
				"normalizedText": normalized_statement(requirement.normalized_text),
				"category": requirement.category.value,
				"suggestedKind": kind.value,
				"suggestedWeight": 2 if kind is SuggestedKind.REQUIRED else 1,
				"sourceModality": requirement.source_modality.value,
				"assessability": requirement.assessability.value,
				"predicate": predicate,
				"evidence": evidence,
				"confidence": min(float(requirement.confidence), 0.8),
				"signals": ["model", "grounded_quote"],
			}
		)
		candidates.append(candidate)
	warnings = list(extraction.warnings)
	if dropped:
		warnings.append(
			f"{dropped} model requirements were omitted because they were unsafe or ungrounded"
		)
	return candidates, warnings


def normalize_model_predicate(predicate: dict[str, object]) -> dict[str, object]:
	vocabulary = load_vocabulary()
	criteria = cast(list[dict[str, object]], predicate["criteria"])
	for criterion in criteria:
		name = criterion.get("canonicalName")
		if criterion.get("type") == CriterionType.SKILL.value and isinstance(name, str):
			criterion["canonicalName"] = vocabulary.phrase_to_canonical.get(
				name.casefold(), name
			)
	return predicate


def deduplicate(
	candidates: Sequence[dict[str, object]], warnings: list[str]
) -> list[dict[str, object]]:
	merged: dict[str, dict[str, object]] = {}
	thresholds: dict[tuple[str, tuple[str, ...]], set[int]] = {}
	for candidate in sorted(
		candidates,
		key=lambda item: float(cast(int | float, item["confidence"])),
		reverse=True,
	):
		key = predicate_key(cast(dict[str, object], candidate["predicate"]), candidate)
		if key in merged:
			existing = merged[key]
			existing["evidence"] = merge_evidence(existing["evidence"], candidate["evidence"])
			existing["signals"] = sorted(
				set(cast(list[str], existing["signals"]))
				| set(cast(list[str], candidate["signals"]))
			)
			existing["confidence"] = min(
				0.98,
				float(cast(int | float, existing["confidence"])) + 0.08,
			)
			continue
		merged[key] = candidate
		predicate = cast(dict[str, object], candidate["predicate"])
		for criterion in cast(list[dict[str, object]], predicate["criteria"]):
			if criterion.get("type") != CriterionType.EXPERIENCE.value:
				continue
			subjects = tuple(sorted(cast(list[str], criterion.get("subjects") or [])))
			threshold = criterion.get("minimumMonths")
			if isinstance(threshold, int):
				thresholds.setdefault(("experience", subjects), set()).add(threshold)
	for (_, subjects), values in thresholds.items():
		if len(values) > 1:
			label = ", ".join(subjects) if subjects else "general experience"
			warnings.append(
				f"Conflicting thresholds found for {label}; recruiter review is required"
			)
	return list(merged.values())


def predicate_key(predicate: dict[str, object], candidate: Mapping[str, object]) -> str:
	criteria = cast(list[dict[str, object]], predicate["criteria"])
	if all(item.get("type") == CriterionType.OTHER.value for item in criteria):
		return "text:" + str(candidate["normalizedText"]).casefold()
	return "predicate:" + json.dumps(predicate, sort_keys=True, separators=(",", ":"))


def merge_evidence(left: object, right: object) -> list[dict[str, object]]:
	entries = cast(list[dict[str, object]], left) + cast(list[dict[str, object]], right)
	seen: set[tuple[str, int, int]] = set()
	merged: list[dict[str, object]] = []
	for entry in entries:
		key = (
			str(entry["blockId"]),
			int(cast(int, entry["startOffset"])),
			int(cast(int, entry["endOffset"])),
		)
		if key not in seen:
			seen.add(key)
			merged.append(entry)
	return merged


def finalize_candidate(candidate: dict[str, object]) -> dict[str, object]:
	evidence = cast(list[dict[str, object]], candidate["evidence"])
	identity = {
		"text": candidate["normalizedText"],
		"predicate": candidate["predicate"],
		"source": [
			(entry["blockId"], entry["startOffset"], entry["endOffset"])
			for entry in evidence
		],
	}
	digest = blake2b(
		json.dumps(identity, sort_keys=True).encode(),
		digest_size=8,
	).hexdigest()
	return {"stableId": f"requirement-{digest}", **candidate}


def education_level(text: str) -> str | None:
	lowered = text.casefold()
	for level, phrases in DEGREE_LEVELS.items():
		if any(phrase in lowered for phrase in phrases):
			return level
	return None


def normalized_statement(text: str) -> str:
	return " ".join(unicodedata.normalize("NFKC", text).split()).strip(" ;")


def unique_strings(values: Sequence[str]) -> list[str]:
	seen: set[str] = set()
	result: list[str] = []
	for value in values:
		key = value.casefold()
		if key not in seen:
			seen.add(key)
			result.append(value)
	return result

File: apps/worker-new/src/job-descriptions/compiler.ts
import { createHash } from "node:crypto";
import { loadVocabulary, mentionedSkills } from "../domain/vocabulary.ts";
import { JOB_REQUIREMENTS_COMPILER_VERSION, JOB_REQUIREMENTS_PROMPT_VERSION, JOB_REQUIREMENTS_SCHEMA_VERSION } from "@skillsignal/server-core/versions";

export const MAX_DRAFT_REQUIREMENTS = 50;

const HEADING_SECTIONS: Record<string, string> = {
	requirements: "requirements",
	qualifications: "requirements",
	"minimum qualifications": "requirements",
	"what you bring": "requirements",
	"what we are looking for": "requirements",
	"what we're looking for": "requirements",
	"preferred qualifications": "preferred",
	"preferred skills": "preferred",
	"nice to have": "preferred",
	"nice-to-have": "preferred",
	"bonus points": "preferred",
	responsibilities: "responsibilities",
	"what you will do": "responsibilities",
	"what you'll do": "responsibilities",
	"about us": "about",
	"about the company": "about",
	benefits: "benefits",
	"what we offer": "benefits",
	"equal opportunity": "legal",
	"how to apply": "application",
};

const REQUIRED_CUE = /\b(must|required|minimum|need to|needs to|shall|essential)\b/i;
const PREFERRED_CUE = /\b(preferred|nice[ -]to[ -]have|bonus|plus|desirable|ideally)\b/i;
const YEARS_PATTERN = /\b(\d{1,2})\s*\+?\s*(?:years?|yrs?)\b/i;
const WORD_YEARS_PATTERN = /\b(one|two|three|four|five|six|seven|eight|nine|ten)\s+years?\b/i;
const NUMBER_WORDS: Record<string, number> = { one: 1, two: 2, three: 3, four: 4, five: 5, six: 6, seven: 7, eight: 8, nine: 9, ten: 10 };
const DEGREE_LEVELS: Record<string, readonly string[]> = {
	doctorate: ["phd", "doctorate", "doctoral", "dphil"],
	master: ["master", "mba", "mtech", "m.tech", "msc", "m.sc"],
	bachelor: ["bachelor", "undergraduate degree", "btech", "b.tech", "bsc", "b.sc"],
};
const PROHIBITED_PATTERN = /\b(age|gender|sex|race|ethnicity|religion|marital status|pregnan(?:t|cy)|disability|nationality|national origin|health|veteran status)\b/i;
const ATTESTATION_PATTERNS: Record<string, RegExp> = {
	work_authorization: /\b(work authorization|authorized to work|visa sponsorship|right to work)\b/i,
	location: /\b(relocat(?:e|ion)|reside in|based in|commutable distance)\b/i,
	schedule: /\b(night shift|weekends?|on[ -]call|travel|time zone|working hours)\b/i,
};
const SOFT_SKILL_PATTERN = /\b(communication|team player|self[ -]starter|leadership style|interpersonal)\b/i;
const CERTIFICATION_PATTERN = /\b(certification|certified|licen[cs]e|credential)\b/i;
const BULLET_PATTERN = /^\s*(?:[-*•▪◦]+|\d+[.)])\s+/;

export type JobBlock = {
	readonly id: string;
	readonly text: string;
	readonly section: string;
	readonly startOffset: number;
	readonly endOffset: number;
	readonly wasBullet: boolean;
};

export const sourceBlocks = (sourceText: string): JobBlock[] => {
	const blocks: JobBlock[] = [];
	let section = "unknown";
	let offset = 0;
	for (const rawLine of sourceText.split(/(?<=\n)/)) {
		const line = rawLine.replace(/\r?\n$/, "");
		const stripped = line.trim();
		const lineStart = offset;
		offset += rawLine.length;
		if (!stripped) continue;
		const headingKey = stripped.replace(/[:\s]+$/, "").toLowerCase();
		if (headingKey in HEADING_SECTIONS && stripped.length <= 80) {
			section = HEADING_SECTIONS[headingKey] as string;
			continue;
		}
		const bullet = BULLET_PATTERN.exec(line);
		const contentStart = bullet ? bullet[0].length : line.length - line.trimStart().length;
		const content = line.slice(contentStart).trim();
		if (content.length < 3) continue;
		const localStart = line.indexOf(content, contentStart);
		const start = lineStart + localStart;
		blocks.push({ id: `jd-b${blocks.length + 1}`, text: content, section, startOffset: start, endOffset: start + content.length, wasBullet: bullet !== null });
	}
	return blocks;
};

export const blocksForModel = (blocks: readonly JobBlock[]): Array<Record<string, string>> =>
	blocks.map((block) => ({ id: block.id, section: block.section, text: block.text }));

export const compileJobDescription = (sourceText: string, modelOutput?: Record<string, unknown> | null, options: { degraded?: boolean; degradedReason?: string | null } = {}): Record<string, unknown> => {
	const blocks = sourceBlocks(sourceText);
	const warnings: string[] = [];
	const candidates: Array<Record<string, unknown>> = deterministicCandidates(blocks, warnings);
	let degraded = options.degraded ?? false;
	if (modelOutput) {
		const [modelCandidates, modelWarnings] = groundedModelCandidates(modelOutput, blocks);
		warnings.push(...modelWarnings);
		candidates.push(...modelCandidates);
		if (!modelCandidates.length) {
			degraded = true;
			warnings.push("Model extraction produced no grounded requirements");
		}
	}
	const requirements = deduplicate(candidates, warnings).slice(0, MAX_DRAFT_REQUIREMENTS);
	if (candidates.length > MAX_DRAFT_REQUIREMENTS) warnings.push("Low-confidence requirements were omitted after the review limit");
	if (options.degradedReason) warnings.push(options.degradedReason);
	const qualityState = requirements.length && !warnings.length && !degraded ? "ready" : "review_required";
	if (!requirements.length) warnings.push("No explicit job requirements were found; add criteria manually");
	return {
		schemaVersion: JOB_REQUIREMENTS_SCHEMA_VERSION,
		compilerVersion: JOB_REQUIREMENTS_COMPILER_VERSION,
		promptVersion: JOB_REQUIREMENTS_PROMPT_VERSION,
		degraded,
		qualityState,
		warnings: uniqueStrings(warnings),
		requirements,
	};
};

const deterministicCandidates = (blocks: readonly JobBlock[], warnings: string[]): Array<Record<string, unknown>> => {
	const candidates: Array<Record<string, unknown>> = [];
	for (const block of blocks) {
		const text = normalizedStatement(block.text);
		if (["about", "benefits", "legal", "application"].includes(block.section)) continue;
		if (PROHIBITED_PATTERN.test(text)) {
			warnings.push(`Potentially prohibited criterion omitted from ${block.id}`);
			continue;
		}
		const skills = [...mentionedSkills(text)].sort((a, b) => a.toLowerCase().localeCompare(b.toLowerCase()));
		const hasCue = REQUIRED_CUE.test(text) || PREFERRED_CUE.test(text);
		const inRequirementSection = block.section === "requirements" || block.section === "preferred";
		if (block.section === "responsibilities" && !hasCue) continue;
		if (!inRequirementSection && !hasCue && !(block.wasBullet && skills.length)) continue;
		candidates.push(deterministicCandidate(block, text, skills));
	}
	return candidates;
};

const deterministicCandidate = (block: JobBlock, text: string, skills: readonly string[]): Record<string, unknown> => {
	const [suggestedKind, sourceModality] = inferImportance(block.section, text);
	const [category, assessability] = inferCategoryAndAssessability(text, skills);
	const predicate = inferPredicate(text, skills);
	const evidence = [{ blockId: block.id, quote: block.text, startOffset: block.startOffset, endOffset: block.endOffset, section: block.section }];
	const signals: string[] = [];
	if (block.section === "requirements" || block.section === "preferred") signals.push("section");
	if (skills.length) signals.push("taxonomy");
	if (REQUIRED_CUE.test(text) || PREFERRED_CUE.test(text)) signals.push("language_cue");
	const confidence = Math.min(0.92, 0.55 + 0.12 * signals.length);
	return finalizeCandidate({
		normalizedText: text,
		category,
		suggestedKind,
		suggestedWeight: suggestedKind === "required" ? 2 : 1,
		sourceModality,
		assessability,
		predicate,
		evidence,
		confidence,
		signals: signals.length ? signals : ["bullet"],
	});
};

const inferImportance = (section: string, text: string): [string, string] => {
	if (PREFERRED_CUE.test(text)) return ["preferred", "explicit_preferred"];
	if (REQUIRED_CUE.test(text)) return ["required", "explicit_required"];
	if (section === "preferred") return ["preferred", "section_preferred"];
	if (section === "requirements") return ["required", "section_required"];
	return ["preferred", "unclear"];
};

const inferCategoryAndAssessability = (text: string, skills: readonly string[]): [string, string] => {
	for (const [category, pattern] of Object.entries(ATTESTATION_PATTERNS)) {
		if (pattern.test(text)) return [category, "candidate_attestation"];
	}
	if (SOFT_SKILL_PATTERN.test(text)) return ["soft_skill", "recruiter_review"];
	if (experienceMonths(text) !== null) return ["experience", "resume_evidence"];
	if (educationLevel(text)) return ["education", "resume_evidence"];
	if (CERTIFICATION_PATTERN.test(text)) return ["certification", "resume_evidence"];
	if (skills.length) return ["skill", "resume_evidence"];
	return ["other", "unclear"];
};

const inferPredicate = (text: string, skills: readonly string[]): Record<string, unknown> => {
	const criteria: Array<Record<string, unknown>> = [];
	const minimumMonths = experienceMonths(text);
	if (minimumMonths !== null) {
		criteria.push({ type: "experience", canonicalName: null, minimumMonths, minimumLevel: null, subjects: [...skills] });
	} else if (skills.length) {
		for (const skill of skills) criteria.push({ type: "skill", canonicalName: skill, minimumMonths: null, minimumLevel: null, subjects: [] });
	}
	const level = educationLevel(text);
	if (level) criteria.push({ type: "education", canonicalName: null, minimumMonths: null, minimumLevel: level, subjects: [] });
	if (CERTIFICATION_PATTERN.test(text) && !criteria.length) {
		criteria.push({ type: "certification", canonicalName: text, minimumMonths: null, minimumLevel: null, subjects: [] });
	}
	if (!criteria.length) criteria.push({ type: "other", canonicalName: null, minimumMonths: null, minimumLevel: null, subjects: [] });
	return { operator: /\bor\b/i.test(text) ? "any_of" : "all_of", criteria };
};

type ModelExtraction = { requirements: Array<Record<string, unknown>>; warnings: string[] };

const groundedModelCandidates = (modelOutput: Record<string, unknown>, blocks: readonly JobBlock[]): [Array<Record<string, unknown>>, string[]] => {
	const byId = new Map(blocks.map((block) => [block.id, block]));
	const requirements = Array.isArray(modelOutput["requirements"]) ? modelOutput["requirements"] as Array<Record<string, unknown>> : null;
	if (!requirements) return [[], ["Model requirement output failed schema validation"]];
	const warnings = Array.isArray(modelOutput["warnings"]) ? (modelOutput["warnings"] as unknown[]).filter((item): item is string => typeof item === "string") : [];
	const candidates: Array<Record<string, unknown>> = [];
	let dropped = 0;
	for (const requirement of requirements) {
		const rawEvidence = Array.isArray(requirement["evidence"]) ? requirement["evidence"] as Array<Record<string, unknown>> : [];
		const evidence: Array<Record<string, unknown>> = [];
		for (const citation of rawEvidence) {
			const block = byId.get(String(citation["blockId"] ?? ""));
			if (!block) continue;
			const quote = String(citation["quote"] ?? "");
			const localStart = block.text.indexOf(quote);
			if (localStart < 0 || !quote) continue;
			evidence.push({ blockId: block.id, quote, startOffset: block.startOffset + localStart, endOffset: block.startOffset + localStart + quote.length, section: block.section });
		}
		if (!evidence.length || requirement["assessability"] === "prohibited") {
			dropped += 1;
			continue;
		}
		const predicate = normalizeModelPredicate((requirement["predicate"] ?? {}) as Record<string, unknown>);
		if (!modelPredicateSupported(predicate, evidence)) {
			dropped += 1;
			continue;
		}
		const kind = String(requirement["suggestedKind"] ?? "preferred");
		candidates.push(finalizeCandidate({
			normalizedText: normalizedStatement(String(requirement["normalizedText"] ?? "")),
			category: String(requirement["category"] ?? "other"),
			suggestedKind: kind,
			suggestedWeight: kind === "required" ? 2 : 1,
			sourceModality: String(requirement["sourceModality"] ?? "unclear"),
			assessability: String(requirement["assessability"] ?? "unclear"),
			predicate,
			evidence,
			confidence: Math.min(Number(requirement["confidence"] ?? 0), 0.8),
			signals: ["model", "grounded_quote"],
		}));
	}
	if (dropped) warnings.push(`${dropped} model requirements were omitted because they were unsafe or ungrounded`);
	return [candidates, warnings];
};

const normalizeModelPredicate = (predicate: Record<string, unknown>): Record<string, unknown> => {
	const vocabulary = loadVocabulary();
	const criteria = Array.isArray(predicate["criteria"]) ? predicate["criteria"] as Array<Record<string, unknown>> : [];
	for (const criterion of criteria) {
		const name = criterion["canonicalName"];
		if (criterion["type"] === "skill" && typeof name === "string") {
			criterion["canonicalName"] = vocabulary.phraseToCanonical.get(name.toLowerCase()) ?? name;
		}
	}
	return { ...predicate, criteria };
};

const modelPredicateSupported = (predicate: Record<string, unknown>, evidence: readonly Record<string, unknown>[]): boolean => {
	const evidenceText = evidence.map((item) => String(item["quote"] ?? "")).join(" ");
	if (PROHIBITED_PATTERN.test(evidenceText)) return false;
	const evidenceSkills = mentionedSkills(evidenceText);
	const criteria = predicate["criteria"];
	if (!Array.isArray(criteria)) return false;
	for (const raw of criteria) {
		if (typeof raw !== "object" || raw === null) return false;
		const criterion = raw as Record<string, unknown>;
		const type = String(criterion["type"] ?? "other");
		if (type === "skill") {
			const name = String(criterion["canonicalName"] ?? "").trim();
			if (!evidenceSkills.has(name) && !evidenceText.toLowerCase().includes(name.toLowerCase())) return false;
		} else if (type === "experience") {
			if (typeof criterion["minimumMonths"] !== "number" || experienceMonths(evidenceText) !== criterion["minimumMonths"]) return false;
			const subjects = criterion["subjects"];
			if (Array.isArray(subjects) && subjects.some((subject) => !evidenceSkills.has(String(subject)) && !evidenceText.toLowerCase().includes(String(subject).toLowerCase()))) return false;
		} else if (type === "education") {
			if (String(criterion["minimumLevel"] ?? "") !== (educationLevel(evidenceText) ?? "")) return false;
		} else if (type === "certification") {
			const name = String(criterion["canonicalName"] ?? "").trim();
			if (!name || !evidenceText.toLowerCase().includes(name.toLowerCase())) return false;
		}
	}
	return true;
};

const deduplicate = (candidates: readonly Record<string, unknown>[], warnings: string[]): Array<Record<string, unknown>> => {
	const merged = new Map<string, Record<string, unknown>>();
	const thresholds = new Map<string, Set<number>>();
	for (const candidate of [...candidates].sort((a, b) => Number(b["confidence"]) - Number(a["confidence"]))) {
		const key = predicateKey(candidate["predicate"] as Record<string, unknown>, candidate);
		const existing = merged.get(key);
		if (existing) {
			existing["evidence"] = mergeEvidence(existing["evidence"], candidate["evidence"]);
			existing["signals"] = [...new Set([...(existing["signals"] as string[]), ...(candidate["signals"] as string[])])].sort();
			existing["confidence"] = Math.min(0.98, Number(existing["confidence"]) + 0.08);
			continue;
		}
		merged.set(key, { ...candidate });
		for (const criterion of ((candidate["predicate"] as Record<string, unknown>)["criteria"] as Array<Record<string, unknown>>)) {
			if (criterion["type"] !== "experience") continue;
			const subjects = JSON.stringify([...((criterion["subjects"] as string[] | null) ?? [])].sort());
			if (typeof criterion["minimumMonths"] === "number") {
				const set = thresholds.get(`experience:${subjects}`) ?? new Set<number>();
				set.add(criterion["minimumMonths"]);
				thresholds.set(`experience:${subjects}`, set);
			}
		}
	}
	for (const [key, values] of thresholds) {
		if (values.size > 1) {
			const subjects = JSON.parse(key.slice("experience:".length)) as string[];
			warnings.push(`Conflicting thresholds found for ${subjects.length ? subjects.join(", ") : "general experience"}; recruiter review is required`);
		}
	}
	return [...merged.values()];
};

const predicateKey = (predicate: Record<string, unknown>, candidate: Record<string, unknown>): string => {
	const criteria = predicate["criteria"] as Array<Record<string, unknown>>;
	if (criteria.every((item) => item["type"] === "other")) return `text:${String(candidate["normalizedText"]).toLowerCase()}`;
	return `predicate:${JSON.stringify(predicate)}`;
};

const mergeEvidence = (left: unknown, right: unknown): Array<Record<string, unknown>> => {
	const entries = [...(left as Array<Record<string, unknown>>), ...(right as Array<Record<string, unknown>>)];
	const seen = new Set<string>();
	const merged: Array<Record<string, unknown>> = [];
	for (const entry of entries) {
		const key = `${String(entry["blockId"])}:${Number(entry["startOffset"])}:${Number(entry["endOffset"])}`;
		if (!seen.has(key)) {
			seen.add(key);
			merged.push(entry);
		}
	}
	return merged;
};

const finalizeCandidate = (candidate: Record<string, unknown>): Record<string, unknown> => {
	const evidence = candidate["evidence"] as Array<Record<string, unknown>>;
	const identity = {
		text: candidate["normalizedText"],
		predicate: candidate["predicate"],
		source: evidence.map((entry) => [entry["blockId"], entry["startOffset"], entry["endOffset"]]),
	};
	const digest = createHash("sha256").update(JSON.stringify(identity), "utf8").digest("hex").slice(0, 16);
	return { stableId: `requirement-${digest}`, ...candidate };
};

export const educationLevel = (text: string): string | null => {
	const lowered = text.toLowerCase();
	for (const [level, phrases] of Object.entries(DEGREE_LEVELS)) {
		if (phrases.some((phrase) => lowered.includes(phrase))) return level;
	}
	return null;
};

export const experienceMonths = (text: string): number | null => {
	const digits = YEARS_PATTERN.exec(text);
	if (digits?.[1]) return Number(digits[1]) * 12;
	const words = WORD_YEARS_PATTERN.exec(text);
	if (words?.[1]) return (NUMBER_WORDS[words[1].toLowerCase()] ?? 0) * 12 || null;
	return null;
};

export const normalizedStatement = (text: string): string =>
	text.normalize("NFKC").split(/\s+/).join(" ").trim().replace(/ ;$/, "").replace(/;$/, "");

const uniqueStrings = (values: readonly string[]): string[] => {
	const seen = new Set<string>();
	const result: string[] = [];
	for (const value of values) {
		if (!seen.has(value.toLowerCase())) {
			seen.add(value.toLowerCase());
			result.push(value);
		}
	}
	return result;
};

export type { ModelExtraction };
